File: src/main/java/org/apache/prepbuddy/filetypes/FileType.java
package org.apache.prepbuddy.filetypes;


import org.apache.commons.lang.StringUtils;

public enum FileType {
    CSV {
        @Override
        public String[] parseRecord(String record) {
            return record.split(",", -1);
        }

        @Override
        public String join(String[] record) {
            return StringUtils.join(record, ",");
        }

        @Override
        public String appendDelimeter(String row) {
            return row + ",";
        }
    },

    TSV {
        @Override
        public String[] parseRecord(String record) {
            return record.split("\t", -1);
        }

        @Override
        public String join(String[] record) {
            return StringUtils.join(record, "\t");
        }

        @Override
        public String appendDelimeter(String row) {
            return row + "\t";
        }
    };

    public abstract String[] parseRecord(String record);

    public abstract String join(String[] record);


    public abstract String appendDelimeter(String row);
}

File: src/main/scala/com/thoughtworks/datacommons/prepbuddy/types/FileType.scala
package com.thoughtworks.datacommons.prepbuddy.types

import com.opencsv.CSVParser
import com.thoughtworks.datacommons.prepbuddy.utils.RowRecord

/**
  * File formats that are supported by TransformableRDD
  */
class FileType(delimiter: String) extends Serializable {
    private val escapeChar = "\""

    def appendDelimiter(row: String): String = row + delimiter

    def join(record: RowRecord): String = {
        record.map((x)=>{
            if(x.contains(delimiter)) escapeChar + x + escapeChar else x
        }).mkString(delimiter)
    }

    def parse(record: String): RowRecord = {
        val csvParser: CSVParser = new CSVParser(delimiter.charAt(0))
        val strings: Array[String] = csvParser.parseLine(record)
        new RowRecord(strings.map(_.trim))
    }
}

object CSV extends FileType(",")

object TSV extends FileType("\t")

